Generic type bounds are not handled

I have a use-case where I would like to wrap the content for individual pages with a generic "page + metadata" content struct (e.g. to provide i18n, script information, etc).

```rust
#[derive(ramhorns::Content)]
struct SamplePage {
    username: String,
    age: u64,
}

#[derive(ramhorns::Content)]
struct PageWithMetadata<T: ramhorns::Content> {
    #[ramhorns(flatten)]
    page: T,
    metadata: (),
}
```

However, this fails with a series of fairly-indecipherable errors:

```
error[E0658]: associated type bounds are unstable
 --> src/main.rs:8:25
  |
8 | struct PageWithMetadata<T: ramhorns::Content> {
  |                         ^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #52662 <https://github.com/rust-lang/rust/issues/52662> for more information

error[E0107]: wrong number of type arguments: expected 1, found 0
 --> src/main.rs:8:8
  |
8 | struct PageWithMetadata<T: ramhorns::Content> {
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 1 type argument

error[E0229]: associated type bindings are not allowed here
 --> src/main.rs:8:25
  |
8 | struct PageWithMetadata<T: ramhorns::Content> {
  |                         ^^^^^^^^^^^^^^^^^^^^ associated type not allowed here
```

Reading briefly through the source, it seems like the generic part is not being parsed or broken down further, and simply being quoted in when necessary, which works for simple type parameters (e.g. `<T>`) but not when there are bounds on those types.

I also tried this by using a `where T: ramhorns::Content` block:

```
error[E0277]: the trait bound `T: ramhorns::Content` is not satisfied
 --> src/main.rs:7:10
  |
7 | #[derive(ramhorns::Content)]
  |          ^^^^^^^^^^^^^^^^^ the trait `ramhorns::Content` is not implemented for `T`
8 | struct PageWithMetadata<T> where T: ramhorns::Content {
  |                                     ----------------- required by this bound in `PageWithMetadata`
  |
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider restricting type parameter `T`
  |
8 | struct PageWithMetadata<T: ramhorns::Content> where T: ramhorns::Content {
  |                          ^^^^^^^^^^^^^^^^^^^
```

Is this supported at all/is there an obvious alternative way to achieve this?  My other way of solving this would be by putting the metadata inside the `SamplePage` struct, but this involves more repetition and messy APIs, so I was hoping to avoid this.

Thanks!

--- a/ramhorns-derive/src/lib.rs
+++ b/ramhorns-derive/src/lib.rs
@@ -66,6 +66,7 @@ pub fn content_derive(input: TokenStream) -> TokenStream {
 
     let name = &item.ident;
     let generics = &item.generics;
+    let type_params = item.generics.type_params();
     let unit_fields = UnitFields::new();
 
     let mut errors = Vec::new();
@@ -222,9 +223,12 @@ pub fn content_derive(input: TokenStream) -> TokenStream {
     let flatten = &*flatten;
     let fields = fields.iter().map(|Field { field, .. }| field);
 
+    let where_clause = type_params.map(|param| quote!(#param: ::ramhorns::Content)).collect::<Vec<_>>();
+    let where_clause = if where_clause.len() > 0 { quote!(where #(#where_clause),*) } else { quote!() };
+
     // FIXME: decouple lifetimes from actual generics with trait boundaries
     let tokens = quote! {
-        impl#generics ramhorns::Content for #name#generics {
+        impl#generics ramhorns::Content for #name#generics #where_clause {
             #[inline]
             fn capacity_hint(&self, tpl: &ramhorns::Template) -> usize {
                 tpl.capacity_hint() #( + self.#fields.capacity_hint(tpl) )*

--- a/tests/tests/main.rs
+++ b/tests/tests/main.rs
@@ -382,6 +382,42 @@ fn can_render_lists_from_vecs() {
 }
 
 #[test]
+fn can_render_generic_types() {
+    #[derive(Content)]
+    struct Article {
+        title: String,
+        body: String,
+    }
+
+    #[derive(Content)]
+    struct Page<T> {
+        #[ramhorns(flatten)]
+        contents: T,
+        last_updated: String,
+    }
+
+    let tpl = Template::new(
+        "<h1>{{title}}</h1>\
+         <article>{{body}}</article>\
+         <p>{{last_updated}}</p>",
+    )
+    .unwrap();
+
+    let html = tpl.render(&Page {
+        last_updated: "yesterday".into(),
+        contents: Article {
+            title: "Jam".into(),
+            body: "This is an article body".into(),
+        },
+    });
+
+    assert_eq!(
+        html,
+        "<h1>Jam</h1><article>This is an article body</article><p>yesterday</p>"
+    )
+}
+
+#[test]
 fn can_render_markdown() {
     #[derive(Content)]
     struct Post<'a> {
@@ -554,13 +590,13 @@ fn struct_with_many_sections() {
     }
 
     let tpl = Template::new("<h1>{{#d}}{{#0}}{{#c}}{{0}}{{/c}}{{/0}}{{/d}} world!</h1>").unwrap();
-    
+
     let rendered = tpl.render(&Page {
         a: A(1),
         b: B(2.0),
         c: C("Hello"),
         d: D(true),
-        other: &[]
+        other: &[],
     });
 
     assert_eq!(rendered, "<h1>Hello world!</h1>");
@@ -606,10 +642,13 @@ fn derive_flatten() {
         title: "This is the title",
         child: Child {
             body: "This is the body",
-        }
+        },
     });
 
-    assert_eq!(html, "<h1>This is the title</h1><head>This is the body</head>");
+    assert_eq!(
+        html,
+        "<h1>This is the title</h1><head>This is the body</head>"
+    );
 }
 
 #[test]

--- a/tests/tests/main.rs
+++ b/tests/tests/main.rs
@@ -382,7 +382,45 @@ fn can_render_lists_from_vecs() {
 }
 
 #[test]
-fn can_render_generic_types() {
+fn can_render_nested_generic_types() {
+    #[derive(Content)]
+    struct Article {
+        title: String,
+        body: String,
+    }
+
+    #[derive(Content)]
+    struct Page<T> {
+        contents: T,
+        last_updated: String,
+    }
+
+    let tpl = Template::new(
+        "\
+        {{#contents}}\
+        <h1>{{title}}</h1>\
+        <article>{{body}}</article>\
+        {{/contents}}\
+        <p>{{last_updated}}</p>",
+    )
+    .unwrap();
+
+    let html = tpl.render(&Page {
+        last_updated: "yesterday".into(),
+        contents: Article {
+            title: "Jam".into(),
+            body: "This is an article body".into(),
+        },
+    });
+
+    assert_eq!(
+        html,
+        "<h1>Jam</h1><article>This is an article body</article><p>yesterday</p>"
+    )
+}
+
+#[test]
+fn can_render_flattened_generic_types() {
     #[derive(Content)]
     struct Article {
         title: String,

